### DISCUSSION TITLE
mastering chain in Jamulus server

### DISCUSSION THREAD

**Discussion by badstorm77** [jamulussoftware/jamulus#3044] [Technical Support & Help]
Hello everyone.
I have been using Jamulus server on my windows pc for almost 3 years. I have no skills as a programmer. i'm a musician.
I ask if it is possible to make a sound card work inside jamulus server so as to be able to apply an analog mastering chain to all the sound of the server. I managed to do this with jactrip server and jack router, but this way I don't have the full jamulus functionality. Do you think my request can be fulfilled?

**Comment by pljones** [comment]
It's probably best to do this by recording at the server and then processing "off line", i.e. not during the session.  Jamulus Server can supply both Audacity (open source, cross-platform) and Reaper (reasonable cost, cross-platform) project files for the recordings of sessions on your server.

I say "best" because you'd be adding latency unnecessarily if you loop through external software or hardware (and Jamulus Server doesn't output or input audio signals anyway).  You'd also have the issue that different players might want to hear things differently (that's why each player in the session can control their own mix).

**Comment by badstorm77** [comment]
I'll try to explain myself better.

I tried using Jacktrip 
( https://github.com/jacktrip/jacktrip/releases )
e
 Jack router ( https://jackaudio.org/downloads/ ).

Jacktrip and JackRT are installed on a windows pc with an ESI MAYA 44 PCIe sound card.

From another mac computer I connected to the server with jacktrip in client mode. 

JackRT takes advantage of the asio drivers of the ESI MAYA 44 card.

Once connected to the server I connected the recive1 and recive2 outputs of the client to outputs 1 and 2 of the card. From physical outputs 1 and 2 I ran the audio signal through an analog mastering chain. Preamp, channelstrip, graphic equalizer, and exciter. From the mastering chain the signal re-enters the physical inputs 1 and 2 of the ESI board. In JackRT the inputs 1 and 2 are capture1 and capture2 which go back into the client. The result is a much deeper and more present sound. Unfortunately, there are not the same audio mixing controls as in jamulus. If it were possible to do the same thing with jamulus server it would be wonderful for live sessions. I hope I have explained myself.

Translated with www.DeepL.com/Translator (free version)
![Schermata 2023-04-13 alle 23 32 15](https://user-images.githubusercontent.com/62301613/231890536-9d8ce30b-782f-4aba-ac0e-50da949cc77f.png)

![Schermata 2023-04-13 alle 23 38 27](https://user-images.githubusercontent.com/62301613/231890593-c383f54c-e7db-4162-b1af-e1eefddddd77.png)

![Schermata 2023-04-13 alle 23 38 44](https://user-images.githubusercontent.com/62301613/231890687-ef079c28-35a3-4434-aaa1-40bbb7f29a7d.png)

![Schermata 2023-04-08 alle 20 36 49](https://user-images.githubusercontent.com/62301613/231890727-8a12479c-b4c9-4b17-be76-e810752e375e.png)

![Schermata 2023-04-13 alle 23 41 12](https://user-images.githubusercontent.com/62301613/231890742-65dde6e1-f570-4d7b-b722-a8af5ccba60e.png)

![WhatsApp Image 2023-04-13 at 23 52 12](https://user-images.githubusercontent.com/62301613/231890938-6ae5779e-39c8-4253-a5ce-6d36dcec6f1b.jpeg)

**Comment by dcorson-ticino-com** [comment]
The Jamulus server is a purely digital mixer.  It has no concept of what a sound card is and would not react well to the added delays introduced by a sound card.  

The server does a different mix for each client connected, there is no place where one could loop out to a common signal processor for all the clients.  This only works because the digital mix is much much faster than the playing time.  Adding an analog step would need a separate analog processor for each connected client. 

Note that you can do client side processing such as you are doing here with Jamulus and then stream the results to your listeners.  Take a look at the WorldJam documentation to see how they do things.

**Comment by ann0see** [comment]
@badstorm77 this is possible by running a client on the same server hardware. It can even be headless - I think at least to some extent.

The issue is that there the server is not capable of outputting any audio (= the Jamulus architecture just doesn't work like that). Maybe we have another way to solve your problem or make mixing on the server via a connected client easier?
Maybe you want to have a mix for ALL users? Then there's the singlemix branch.
